### PR TITLE
only send eds for updated services 

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

Service、endpoint are the most frequently updated resources that trigger xds push.

We only need to push eds for the updated clusters, here we calc it based on configsUpdated only contains ServiceEntry.

### CODE REVIEW

File: operator/cmd/mesh/install.go [hunk 1]
```diff
@@ -247,7 +247,13 @@ func InstallManifests(iop *v1alpha12.IstioOperator, force bool, dryRun bool, kub
 		return iop, fmt.Errorf("errors occurred during operation: %v", err)
 	}
 	if status.Status != v1alpha1.InstallStatus_HEALTHY {
-		return iop, fmt.Errorf("errors occurred during operation")
+		additionalInfo := ""
```
Comment (ramaraochavali) [inline operator/cmd/mesh/install.go:250]: This seems unrelated to this PR

File: pilot/pkg/xds/eds.go [hunk 2]
```diff
@@ -379,28 +379,35 @@ func edsNeedsPush(updates model.XdsUpdates) bool {
...
+
 func (eds *EdsGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource,
 	req *model.PushRequest) (model.Resources, model.XdsLogDetails, error) {
 	if !edsNeedsPush(req.ConfigsUpdated) {
 		return nil, model.DefaultXdsLogDetails, nil
 	}
 	var edsUpdatedServices map[string]struct{}
-	if !req.Full {
+	if !req.Full ||
+		// all configs updated are ServiceEntry type, then we only need to send eds for these updated services.
```
Comment (ramaraochavali) [inline pilot/pkg/xds/eds.go:398]: When will we need this check - it seems if it is not a full push we are only sending eds?

File: operator/cmd/mesh/install.go (hunk 1 above)
Comment (hzxuzhonghu) [reply operator/cmd/mesh/install.go:250]: Yes, i used it to know why the install failed

File: pilot/pkg/xds/eds.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/xds/eds.go:398]: I want to calculate which eds need to push when se updated.  Envoy does not re-warm clusters that are not changed, so we donot need to push eds for them too?

Comment (ramaraochavali) [reply pilot/pkg/xds/eds.go:398]: > I want to calculate which eds need to push when se updated

Where are you doing that? edsUpdatedServices has list of all ses?

Comment (hzxuzhonghu) [reply pilot/pkg/xds/eds.go:398]: Get all the `updated services` and  

ignore those eds for unchanged clusters: w.ResourceNames - `updated services`

Comment (ramaraochavali) [reply pilot/pkg/xds/eds.go:398]: sorry. my question was not clear - it did not change with this PR though, we were already doing that, wondering what is the use of this new condition? 

Are you saying this will be used if req is Full and ConfigUpdated has only serviceEntry update even then you want to send eds only?

Not sure if that correct behaviour because if service is changed should n't we send updated cluster?

Comment (hzxuzhonghu) [reply pilot/pkg/xds/eds.go:398]: This is working on full push

Comment (hzxuzhonghu) [reply pilot/pkg/xds/eds.go:398]: If svc is changed, it will be added to `edsUpdatedServices`

and i meant to gen eds for it

File: pilot/pkg/xds/eds.go [hunk 3]
```diff
@@ -387,7 +396,12 @@ func (eds *EdsGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w
 	var edsUpdatedServices map[string]struct{}
 	if !req.Full {
 		edsUpdatedServices = model.ConfigNamesOfKind(req.ConfigsUpdated, gvk.ServiceEntry)
+	} else if len(req.ConfigsUpdated) > 0 && allConfigsWithKind(req.ConfigsUpdated, gvk.ServiceEntry.Kind) {
+		// when full push
+		// if all configs updated are ServiceEntry type, then we only need to send eds for these updated services.
+		edsUpdatedServices = model.ConfigNamesOfKind(req.ConfigsUpdated, gvk.ServiceEntry)
```
Comment (howardjohn) [inline pilot/pkg/xds/eds.go:402]: What about multinetwork Gateways? is that a concern?

Comment (hzxuzhonghu) [reply pilot/pkg/xds/eds.go:402]: Should be no. Gateway acts the same as sidecars.

**Comment by hzxuzhonghu** [comment]
After some investigation, the failure seems random. And I filed an issue https://github.com/envoyproxy/envoy/issues/19262

**Comment by hzxuzhonghu** [comment]
This is currently not right given to envoy bug https://github.com/envoyproxy/envoy/issues/19262
